File: 20220531/app/src/test/java/SolutionTest.java
import org.junit.jupiter.api.Test;

import java.util.HashSet;
import java.util.stream.IntStream;

import static org.assertj.core.api.Assertions.assertThat;

class SolutionTest {

    private int solution(int[] ponkemons) {
        HashSet<Integer> type = new HashSet<>();
        int amounts = ponkemons.length / 2;

        IntStream.of(ponkemons).forEach(monster -> type.add(monster));

        if (type.size() > amounts) {
            return amounts;
        }
        else {
            return type.size();
        }
    }

    @Test
    void solutionTest() {
        assertThat(solution(new int[]{3, 1, 2, 3})).isEqualTo(2);
        assertThat(solution(new int[]{3, 3, 3, 2, 2, 4})).isEqualTo(3);
        assertThat(solution(new int[]{3, 3, 3, 2, 2, 2})).isEqualTo(2);
    }

}

File: 20250210/kotlin/src/test/kotlin/com/example/kotlin/SolutionTest.kt
package com.example.kotlin

import org.junit.jupiter.api.Test
import org.assertj.core.api.Assertions.assertThat

class SolutionTest {
    @Test
    fun `경주가 끝났을 때 해설진이 이름을 부르는 순서`() {
        assertThat(
            solution(
                arrayOf("mumu", "soe", "poe"),
                arrayOf("soe")
            )
        ).isEqualTo(arrayOf("soe", "mumu", "poe"))
    }

    fun solution(
        players: Array<String>,
        callings: Array<String>,
    ): Array<String> {
        val rank = HashMap<String, Int>()

        players.forEachIndexed { index, player ->
            rank[player] = index
        }

        callings.forEach { player ->
            swap(rank, player, players)
        }

        return players
    }

    fun swap(
        rank: HashMap<String, Int>,
        player: String,
        players: Array<String>
    ) {
        val currentRank = rank[player] ?: return

        if (currentRank == 0) return

        val frontPlayerRank = currentRank - 1
        val frontPlayer = players[frontPlayerRank]

        players[frontPlayerRank] = player
        players[currentRank] = frontPlayer

        rank[player] = frontPlayerRank
        rank[frontPlayer] = currentRank
    }
}
